syntax = "proto3";
option cc_generic_services = true;
option java_generic_services = true;

import "common.proto";

package challenge; 

message StageLayer
{
    repeated boolean stages = 1;
}

enum PointsLayerType
{
    TOTAL = 1;
    WEEKLY = 2;
}

message PointsLayer
{
    uint32 numPoints = 1;
    uint32 pointsToNextLevel = 2;
    uint32 advancementValue = 3;
    PointsLayerType type = 4;
}

message LayerId
{
    common.UniqueId a = 1;
    common.UniqueId b = 2;
}

message Layer
{
    LayerId id = 1;
    StageLayer stage_layer = 2;
    PointsLayer points_layer = 3;
}

message Item
{
    common.UniqueId id = 1;
    string name = 2;
    uint32 level = 3;
    uint32 maxLevels = 4;

    repeated Layer layers = 5;
}

message ChallengeItems
{
    repeated Item items = 1;
}

message AddChallengeParams
{
    string name = 1;
    uint32 maxLevels = 2;
}

message AddLayerParams
{
    common.UniqueId parentId = 1;
    Layer layer = 2;
}

message ToggleStageParams
{
    LayerId id = 1;
    uint32 ordinal_number = 2;
}

message AddPointsParams
{
    LayerId id = 1;
    uint32 points = 2;
}

service ChallengeService 
{
   rpc GetItems (common.EmptyMessage) returns (ChallengeItems);
   rpc DeleteItem (common.UniqueId) returns (common.OperationResultMessage);
   rpc AddItem (AddChallengeParams) returns (common.UniqueId);

   rpc DeleteLayer (LayerId) returns (common.OperationResultMessage);
   rpc AddLayer (AddLayerParams) returns (LayerId);

   rpc ToggleStage(ToggleStageParams) returns (common.OperationResultMessage);
   rpc AddPoints(AddPointsParams) returns (common.OperationResultMessage);
}